data/src/merkle_tree.rs

```diff
@@ -0,0 +1,67 @@
+use crate::hash::Hash;
+
+/// A variable-width Merkle tree with access metadata for leaves.
+///
+/// Values of this type are produced by the proof-generating backend to capture
+/// a snapshot of the machine state along with access information for leaves
+/// which hold data that was used in a particular evaluation step.
+#[derive(Debug, Clone)]
+pub enum MerkleTree {
+    Leaf(Hash, bool, Vec<u8>),
+    Node(Hash, Vec<Self>),
+}
+
+impl MerkleTree {
+    /// Get the root hash of a Merkle tree
+    pub fn root_hash(&self) -> Hash {
+        match self {
+            Self::Node(hash, _) => *hash,
+            Self::Leaf(hash, _, _) => *hash,
+        }
+    }
+
+    /// Make a Merkle tree consisting of a single leaf; representing
+    /// the given data and its access pattern.
+    pub fn make_merkle_leaf(data: Vec<u8>, access_info: bool) -> Self {
+        let hash = Hash::blake3_hash_bytes(&data);
+        MerkleTree::Leaf(hash, access_info, data)
+    }
+
+    /// Make a Merkle tree consisting of a node which combines the given
+    /// child trees.
+    pub fn make_merkle_node(children: Vec<Self>) -> Self {
+        let children_hashes = children.iter().map(|t| t.root_hash());
+        let node_hash = Hash::combine(children_hashes);
+        MerkleTree::Node(node_hash, children)
+    }
+}
+
+impl MerkleTree {
+    /// Check the validity of the Merkle root by recomputing all hashes
+    #[cfg(feature = "for_testing")]
+    pub fn check_root_hash(&self) -> bool {
```

Comment:
I think it's fine to expose this as-is. No need to tie this to a feature flag.

```suggestion
    pub fn check_root_hash(&self) -> bool {
```

Reply:
You have decoupled the definition and usage context now. That means making it test-only again comes with unnecessary complexity and added compilation time as the crate is compiled twice now - once for each feature set.